### DISCUSSION TITLE
Flatpak version not working correctly

### DISCUSSION THREAD

**Discussion by ademlabs** [LizardByte/community#221] [Sunshine]
Hi all, I'm having some difficulty tracking down the issue with the latest Flatpak version, however the normal binaries work fine. I'm not sure what the cause is and I'm using the same version for both. When I use the Flatpak version, moonlight clients are unable to start an application. They see the sunshine instance correctly and even list any new applications that I add to the list, as well as successfully pair with PIN, but unable to start any applications. The sunshine web dashboard works fine as well but whenever I start an application from a client I get the following error messages:

*Moonlight for Android:*
```
**Connection Error**
Failed to start Steam BigPicture (error 0)
Check your firewall and port forwarding rules for port(s):
TCP 47984
TCP 47989
```

*Moonlight for PC:*
```
GeForce Experience returned error: Malformed GFE XML (missing root element) (Error -1)
```

I'm using Flatpak version 0.19.1 on EndeavourOS (Arch), custom kernel 6.2.11-273-tkg-bmq
Running on an AMD 6800XT reference card.
Moonlight clients tried: Android (both mobile and Shield TV) v11.0, PC version 4.3.1

I've followed the post-install instructions for the Flatpak version as well and read through the docs, troubleshooting docs and even looked at the GitHub repository issues to no avail. I'm not sure whether this is a bug yet or maybe some permission issues related to the Flatpak version.
The binary version works perfectly fine though, I'm able to stream anything. The logs don't show anything useful either, no errors.

Running `netstat -taupe | grep -i sunshine` shows that the ports are open:
```
tcp        0      0 0.0.0.0:48010           0.0.0.0:*               LISTEN      arvind     1521465    49266/sunshine
tcp        0      0 0.0.0.0:47990           0.0.0.0:*               LISTEN      arvind     1523746    49266/sunshine
tcp        0      0 0.0.0.0:47989           0.0.0.0:*               LISTEN      arvind     1521467    49266/sunshine
tcp        0      0 0.0.0.0:47984           0.0.0.0:*               LISTEN      arvind     1512269    49266/sunshine
```

**Logs**
Flatpak logs: https://pastebin.com/raw/JgDT90L7
Binary logs: https://pastebin.com/raw/uAaFqZGN

Anyone knows what might be going on?

**Comment by Thaurin** [comment]
I am experiencing the same issue. I am using the Sunshine 0.20.0 AppImage on SteamOS. Connecting with Moonlight from a Windows 11 laptop gives me the same error as you:

    GeForce Experience returned error: Malformed GFE XML (missing root element) (Error -1)

Connecting from iPhone just gives me a generic "Connection Error" popup.

Everything seems to work, except opening an application. The log is not helpful:

    [2023:06:02:10:07:30]: Info: // Testing for available encoders, this may generate errors. You can safely ignore those errors. //
    [2023:06:02:10:07:30]: Info: Trying encoder [nvenc]
    [2023:06:02:10:07:30]: Info: Screencasting with X11
    Cannot load libcuda.so.1
    [2023:06:02:10:07:30]: Error: Couldn't load cuda: -1
    [2023:06:02:10:07:30]: Info: Screencasting with X11
    Cannot load libcuda.so.1
    [2023:06:02:10:07:30]: Error: Couldn't load cuda: -1
    [2023:06:02:10:07:30]: Info: Encoder [nvenc] failed
    [2023:06:02:10:07:30]: Info: Trying encoder [vaapi]
    [2023:06:02:10:07:30]: Info: Screencasting with X11
    [2023:06:02:10:07:30]: Error: Couldn't open EGL display: [00003000]
    [2023:06:02:10:07:30]: Info: Screencasting with X11
    [2023:06:02:10:07:30]: Error: Couldn't open EGL display: [00003000]
    [2023:06:02:10:07:30]: Info: Encoder [vaapi] failed
    [2023:06:02:10:07:30]: Info: Trying encoder [software]
    [2023:06:02:10:07:30]: Info: Screencasting with X11
    [2023:06:02:10:07:30]: Info: SDR color coding [Rec. 601]
    [2023:06:02:10:07:30]: Info: Color range: [JPEG]
    [2023:06:02:10:07:30]: Info: 
    [2023:06:02:10:07:30]: Info: // Ignore any errors mentioned above, they are not relevant. //
    [2023:06:02:10:07:30]: Info: 
    [2023:06:02:10:07:30]: Info: Found encoder software: [libx264]

Is there any more information or tests that I can perform? I'm drawing a blank here.

**Comment by ReenigneArcher** [comment]
I believe this is the same issue as https://github.com/LizardByte/Sunshine/issues/1295

**Comment by Thaurin** [comment]
Is it known which version may have introduced this? I've been trying older AppImage builds, but they all give me the same thing. I originally thought it might be caused on the side of Moonlight on Windows 11, but the iOS app does the same thing (I assume, as the error message is non-descriptive).

**Comment by Thaurin** [comment]
So  I got it to work on my end. I played around with the earliest AppImages available. One of them must've reinserted an application entry for desktop. Earlier, I only had low-res desktop and Steam listed. These stlil don't work. But I went back to 0.20.0 and deleted my `~/.config/sunshine` directory, repaired, and selected the first desktop entry, which worked.

Maybe the config got corrupted by an earlier version, maybe I accidentally deleted it, I don't know. Low-res desktop and Steam still give the same error message, though.

**Reply by ReenigneArcher** [reply]
If you want anyone to help, you need to provide logs. Should probably also open your own thread since AppImage issues with opening apps are unrelated to the same issue in the Flatpak sandbox.

**Reply by Thaurin** [reply]
I had previously provided a partial log, and I did have similar issues with the flatpak, but I understand and I'll leave it at this.

**Comment by ademlabs** [comment]
I still haven't been able to get my flatpak working. I am still running the package binary unfortunately and the answer that I got from Discord was to just use the package binary. Since the project provides a flatpak and appimage, it would be nice to support those to ensure they work, or remove them from the release.

**Reply by ReenigneArcher** [reply]
Flatpak works for many, but not everyone. Similarly some with issues with the deb/rpm packages have success with flatpak.

**Reply by ademlabs** [reply]
Any way to debug this? The logs don't show anything useful. 
I just tried the latest flatpak release and have the same issues where the sunshine server is detected, the pairing works etc but applications cannot start. The unjailed versions work fine though.

**Reply by matthewddunlap** [reply]
@ademlabs 

Are you able to successfully launch the "Desktop" from a client when using Flatpak server?

I just set up the Flatpak server and found that Desktop worked, but applications did not. I was able to work around that by using `flatpak-spawn --host` to stert/stop a custom user systemd service that exec/kill the desired GUI application.

It's quite the round about way to accomplish this, but all I found that worked.

**Reply by ReenigneArcher** [reply]
@matthewddunlap that's why we put that in our docs. https://docs.lizardbyte.dev/projects/sunshine/latest/md_docs_2getting__started.html#considerations

flatpaks run in a sandboxed environment... The `--host` arg is required for flatpak to access applications outside of the flatpak sandbox. https://www.systutorials.com/docs/linux/man/1-flatpak-spawn/

**Comment by zaggynl** [comment]
Hello, I'm trying to launch some Steam games via Moonlight/Sunshine flatpak but have only managed to share my desktop so far, some games work within this.

If I try to run a game directly with:

flatpak-spawn --host steam://rungameid/2321470 
(for deep rock galactic:survivor) 
I end up with a message: ' host returned error: malformed xml (missing root element)  (error -1)'

Edit1
So the command was missing some bits and I did not have flatpak-spawn installed.
After making the command: 'flatpak-spawn --host setsid steam steam://rungameid/2321470' the same xml error still shows, if I run the command from a terminal it works.

**Comment by k-e-i-z-a-i** [comment]
@ademlabs @zaggynl @Thaurin @matthewddunlap

For those unable to get Sunshine to launch apps (such as Steam Games or even into Big Picture mode) and are only able to get Sunshine to launch and show the Desktop, **install [Flatseal](https://github.com/tchx84/Flatseal) (likely available in your app store) to manage Sunshine's permissions and enable its permission for "D-Bus session bus"** and Sunshine will then run as intended (so long as you follow the Sunshine app setup rules as seen here: https://docs.lizardbyte.dev/projects/sunshine/en/latest/about/guides/app_examples.html)

LizardByte might want to take note of this and integrate this fix to address this problem.

**Comment by nycki93** [comment]
In my case I also had to use flatpak-spawn --host, so:

```
# detached command
flatpak-spawn --host steam steam://open/gamepadui
```

**Reply by ReenigneArcher** [reply]
https://docs.lizardbyte.dev/projects/sunshine/latest/md_docs_2getting__started.html#considerations

**Reply by nycki93** [reply]
Fair enough! This would have been easier to find if it was under the Flatpak section.

[a comment by an automated account is not shown]
